common/ferrostar/src/algorithms.rs

```diff
@@ -262,39 +263,43 @@ fn distance_along(point: &Point, linestring: &LineString) -> Option<f64> {
 
 /// Computes the distance between a location and the end of the current route step.
 /// We assume that input location is pre-snapped to route step's linestring.
-fn distance_to_end_of_step(snapped_location: &Point, current_step_linestring: &LineString) -> f64 {
+///
+/// The result may be [None] in case of invalid input such as infinite floats.
+fn distance_to_end_of_step(snapped_location: &Point, current_step_linestring: &LineString) -> Option<f64> {
     let step_length = current_step_linestring.haversine_length();
-    if let Some(traversed) = distance_along(snapped_location, current_step_linestring) {
+    distance_along(snapped_location, current_step_linestring).map(|traversed| {
         step_length - traversed
-    } else {
-        0.0
-    }
+    })
 }
 
 /// Computes the arrival state for a snapped location along the route.
 /// This includes distances and durations.
+///
+/// NOTE: `remaining_steps` includes the current step.
 pub fn calculate_trip_progress(
     snapped_location: &Point,
-    current_step: &RouteStep,
     current_step_linestring: &LineString,
     remaining_steps: &[RouteStep],
 ) -> TripProgress {
-    if remaining_steps.is_empty() {
+    let Some(current_step) = remaining_steps.first() else {
         return TripProgress {
             distance_to_next_maneuver: 0.0,
             distance_remaining: 0.0,
             duration_remaining: 0.0,
         };
-    }
+    };
 
     // Calculate the distance and duration till the end of the current route step.
     let distance_to_next_maneuver =
-        distance_to_end_of_step(snapped_location, current_step_linestring);
+        distance_to_end_of_step(snapped_location, current_step_linestring).unwrap_or(current_step.distance);
 
     // This could be improved with live traffic data along the route.
     // TODO: Figure out the best way to enable this use case
-    let pct_remaining_current_step =
-        distance_to_next_maneuver / current_step_linestring.haversine_length();
+    let pct_remaining_current_step = if current_step.distance > 0f64 {
+        distance_to_next_maneuver / current_step.distance
+    } else {
+        0f64
+    };
```

Comment:
This was the core issue. If the snapped point is exactly at the end of the step, this results in division by zero.